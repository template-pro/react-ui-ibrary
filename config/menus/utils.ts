const utilsMenu = [
  {
    title: 'General',
    children: ['src/with-default-props'],
  },
  {
    title: 'React Hooks',
    children: ['src/use-props-value'],
  },
  // {
  //   title: "Date & Time",
  //   children: [],
  // },
  // {
  //   title: "String",
  //   children: [],
  // },
]

export default utilsMenu
