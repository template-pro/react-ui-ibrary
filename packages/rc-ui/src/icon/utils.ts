import { useEffect } from 'react'
import { insertCss } from 'insert-css'

export const svgBaseProps = {
  // width: '1em',
  // height: '1em',
  'fill': 'currentColor',
  'aria-hidden': 'true',
  'focusable': 'false',
}

export const iconStyles = `
.template-pro-icon {
  display: inline-block;
  width: 24px;
  color: inherit;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.template-pro-icon > * {
  line-height: 1;
}

.template-pro-icon svg {
  display: inline-block;
}

.template-pro-icon::before {
  display: none;
}

.template-pro-icon .template-pro-icon {
  display: block;
}

.template-pro-icon[tabindex] {
  cursor: pointer;
}

.template-pro-icon-spin::before,
.template-pro-icon-spin {
  display: inline-block;
  -webkit-animation: loadingCircle 1s infinite linear;
  animation: loadingCircle 1s infinite linear;
}

@-webkit-keyframes loadingCircle {
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes loadingCircle {
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
`

let cssInjectedFlag = false

export const useInsertStyles = (styleStr = iconStyles) => {
  useEffect(() => {
    if (!cssInjectedFlag) {
      insertCss(styleStr, {
        prepend: true,
      })
      cssInjectedFlag = true
    }
  }, [])
}
