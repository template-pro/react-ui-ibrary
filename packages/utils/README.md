# @template-pro/utils

统一工具库，通用正则，React Hooks 等工具。

## 安装

```bash
npm install @template-pro/utils
```
