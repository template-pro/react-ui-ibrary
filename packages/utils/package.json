{
  "name": "@template-pro/utils",
  "version": "0.0.0",
  "description": "统一工具库，通用正则，react hooks 等工具",
  "main": "./lib/index.js",
  "module": "./es/index.js",
  "types": "./es/index.d.ts",
  "files": [
    "lib",
    "es"
  ],
  "scripts": {
    "build": "gulp",
    "release":"bumpp --tag @template-pro/utils@%s --all --no-verify && pnpm publish"
  },
  "keywords": [
    "utils"
  ],
  "author": "Wuxh<wxh1220@gmail.com>",
  "license": "MIT",
  "publishConfig": {
    "registry": "http://nas.wxhboy.cn:98/"
  }
}
