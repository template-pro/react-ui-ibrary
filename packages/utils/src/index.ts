export { default as usePropsValue } from './use-props-value'
export { default as mergeProps } from './with-default-props'

// RegExp
export * as anyRule from './any-rule'
