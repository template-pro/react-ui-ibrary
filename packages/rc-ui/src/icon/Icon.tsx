import React from 'react'
import classNames from 'classnames'

import type { CustomIconComponentProps, IconProps } from './typings'
import { svgBaseProps, useInsertStyles } from './utils'

const Icon = (props: IconProps) => {
  const {
    className,
    component: Component,
    viewBox,
    spin,
    rotate,
    tabIndex,
    onClick,
    children,
    prefixCls = 'template-pro',
    ...restProps
  } = props

  useInsertStyles()

  const classString = classNames(`${prefixCls}-icon`, className)

  const svgClassString = classNames({
    [`${prefixCls}-icon-spin`]: !!spin,
  })

  const svgStyle = rotate
    ? {
        msTransform: `rotate(${rotate}deg)`,
        transform: `rotate(${rotate}deg)`,
      }
    : undefined

  const innerSvgProps: CustomIconComponentProps = {
    ...svgBaseProps,
    className: svgClassString,
    style: svgStyle,
    viewBox,
  }

  if (!viewBox)
    delete innerSvgProps.viewBox

  const renderInnerNode = (function () {
    if (Component)
      return <Component {...innerSvgProps}>{children}</Component>

    if (children) {
      return (
        <svg {...innerSvgProps} viewBox={viewBox}>
          {children}
        </svg>
      )
    }
    return null
  }())

  let iconTabIndex = tabIndex
  if (iconTabIndex === undefined && onClick)
    iconTabIndex = -1

  return (
    <span
      role="img"
      {...restProps}
      tabIndex={iconTabIndex}
      onClick={onClick}
      className={classString}
    >
      {renderInnerNode}
    </span>
  )
}

Icon.displayName = 'TemplateProIcon'

export default Icon
