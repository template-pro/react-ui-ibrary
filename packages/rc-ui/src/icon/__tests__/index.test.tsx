import React from 'react';
import { render, fireEvent } from '../../../../../tests/utils';

import Icon from '..';

describe('Icon', () => {
  it('should support pass svg paths as children', () => {
    const { container } = render(
      <Icon viewBox="0 0 24 24">
        <title>Cool Home</title>
        <path d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z" />
      </Icon>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should support event listeners when passing svg paths as children', () => {
    const onClickHandler = jest.fn();
    const onKeyUpHandler = jest.fn();
    const onMouseEnterHandler = jest.fn();

    const { container, getByRole } = render(
      <Icon
        viewBox="0 0 24 24"
        onClick={onClickHandler}
        onKeyUp={onKeyUpHandler}
        onMouseEnter={onMouseEnterHandler}
      >
        <title>Cool Home</title>
        <path d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z" />
      </Icon>,
    );
    expect(container).toMatchSnapshot();

    const icon = getByRole('img') as HTMLSpanElement;
    fireEvent.click(icon);
    expect(onClickHandler).toBeCalledTimes(1);
    fireEvent.mouseEnter(icon);
    expect(onMouseEnterHandler).toBeCalledTimes(1);
    fireEvent.keyUp(icon);
    expect(onKeyUpHandler).toBeCalledTimes(1);
  });

  it('should support custom usage of children', () => {
    expect(() => {
      render(<Icon>&E648</Icon>);
    }).not.toThrow();
  });

  it('support render svg as component', () => {
    const renderSvg = () => (
      <svg viewBox="0 0 1024 1024" width="1em" height="1em" fill="currentColor" />
    );
    const SvgIcon = (props: any) => <Icon component={renderSvg} {...props} />;

    expect(render(<SvgIcon />).container).toMatchSnapshot();
  });

  describe('`component` prop', () => {
    it('can access to svg defs if has children', () => {
      const { container } = render(
        <Icon
          className="my-home-icon"
          component={(svgProps) => (
            <svg {...svgProps}>
              <defs>
                <linearGradient id="gradient">
                  <stop offset="20%" stopColor="#39F" />
                  <stop offset="90%" stopColor="#F3F" />
                </linearGradient>
              </defs>
              {React.Children.map(svgProps.children, (child: any) =>
                React.cloneElement(
                  child,
                  child.type === 'path' ? { fill: 'scriptUrl(#gradient)' } : {},
                ),
              )}
            </svg>
          )}
        >
          <title>Cool Home</title>
          <path d="'M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z'" />
        </Icon>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('should support svg react component', () => {
    // children props would make no sense
    const SvgComponent = (props: any) => (
      <svg viewBox="0 0 24 24" {...props}>
        <title>Custom Svg</title>
        <path d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z" />
      </svg>
    );

    const { container } = render(
      <Icon className="my-home-icon" component={SvgComponent}>
        <title>Cool Home</title>
        <path d="'M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z'" />
      </Icon>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should support event listeners', () => {
    const onClickHandler = jest.fn();
    const onKeyUpHandler = jest.fn();
    const onMouseEnterHandler = jest.fn();
    // children props would make no sense
    const SvgComponent = (props: any) => (
      <svg viewBox="0 0 24 24" {...props}>
        <title>Custom Svg</title>
        <path d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z" />
      </svg>
    );

    const { container, getByRole } = render(
      <Icon
        className="my-home-icon"
        component={SvgComponent}
        onClick={onClickHandler}
        onKeyUp={onKeyUpHandler}
        onMouseEnter={onMouseEnterHandler}
      >
        <title>Cool Home</title>
        <path d="'M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z'" />
      </Icon>,
    );
    expect(container).toMatchSnapshot();

    const icon = getByRole('img') as HTMLSpanElement;
    fireEvent.click(icon);
    expect(onClickHandler).toBeCalledTimes(1);
    fireEvent.mouseEnter(icon);
    expect(onMouseEnterHandler).toBeCalledTimes(1);
    fireEvent.keyUp(icon);
    expect(onKeyUpHandler).toBeCalledTimes(1);
  });
});

describe('Render with styles', () => {
  it('icon style will inset top of head', () => {
    const head = document.querySelector('head')!;
    const meta = document.createElement('meta')!;
    head.appendChild(meta);
    render(<Icon component={'Antd' as any} />);
    expect(head.firstElementChild?.tagName).toBe('STYLE');
  })
});