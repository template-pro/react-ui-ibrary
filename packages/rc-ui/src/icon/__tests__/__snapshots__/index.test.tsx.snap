// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon \`component\` prop can access to svg defs if has children 1`] = `
<div>
  <span
    class="template-pro-icon my-home-icon"
    role="img"
  >
    <svg
      aria-hidden="true"
      class=""
      fill="currentColor"
      focusable="false"
    >
      <defs>
        <lineargradient
          id="gradient"
        >
          <stop
            offset="20%"
            stop-color="#39F"
          />
          <stop
            offset="90%"
            stop-color="#F3F"
          />
        </lineargradient>
      </defs>
      <title>
        Cool Home
      </title>
      <path
        d="'M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z'"
        fill="scriptUrl(#gradient)"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon should support event listeners 1`] = `
<div>
  <span
    class="template-pro-icon my-home-icon"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <title>
        Custom Svg
      </title>
      <path
        d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon should support event listeners when passing svg paths as children 1`] = `
<div>
  <span
    class="template-pro-icon"
    role="img"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <title>
        Cool Home
      </title>
      <path
        d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon should support pass svg paths as children 1`] = `
<div>
  <span
    class="template-pro-icon"
    role="img"
  >
    <svg
      aria-hidden="true"
      class=""
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <title>
        Cool Home
      </title>
      <path
        d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon should support svg react component 1`] = `
<div>
  <span
    class="template-pro-icon my-home-icon"
    role="img"
  >
    <svg
      aria-hidden="true"
      class=""
      fill="currentColor"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <title>
        Custom Svg
      </title>
      <path
        d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
      />
    </svg>
  </span>
</div>
`;

exports[`Icon support render svg as component 1`] = `
<div>
  <span
    class="template-pro-icon"
    role="img"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 1024 1024"
      width="1em"
    />
  </span>
</div>
`;
